linera-views/src/indexed_db.rs

```diff
@@ -0,0 +1,359 @@
+            &lower.into(),
+            &upper.into(),
+            false,
+            true,
+        )
+    } else {
+        web_sys::IdbKeyRange::lower_bound(&lower.into())
+    }
+}
+
+impl LocalReadableKeyValueStore<IndexedDbContextError> for IndexedDbStore {
+    const MAX_KEY_SIZE: usize = usize::MAX;
+    type Keys = Vec<Vec<u8>>;
+    type KeyValues = Vec<(Vec<u8>, Vec<u8>)>;
+
+    fn max_stream_queries(&self) -> usize {
+        self.max_stream_queries
+    }
+
+    async fn read_value_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, IndexedDbContextError> {
+        let key = js_sys::Uint8Array::from(key);
+        let value = self.with_object_store(|o| o.get(&key))??.await?;
+        Ok(value.map(|v| js_sys::Uint8Array::new(&v).to_vec()))
+    }
+
+    async fn contains_key(&self, key: &[u8]) -> Result<bool, IndexedDbContextError> {
+        let key = js_sys::Uint8Array::from(key);
+        let count = self.with_object_store(|o| o.count_with_key(&key))??.await?;
+        assert!(count < 2);
+        Ok(count == 1)
+    }
+
+    async fn read_multi_values_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, IndexedDbContextError> {
+        let mut values = Vec::with_capacity(keys.len());
+        for key in keys {
+            values.push(self.read_value_bytes(&key).await?);
+        }
+        Ok(values)
+    }
+
+    async fn find_keys_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Vec<Vec<u8>>, IndexedDbContextError> {
+        let range = prefix_to_range(key_prefix)?;
+        Ok(self
+            .with_object_store(|o| o.get_all_keys_with_key(&range))??
+            .await?
+            .into_iter()
+            .map(|key| {
+                let key = js_sys::Uint8Array::new(&key);
+                key.subarray(key_prefix.len() as u32, key.length()).to_vec()
+            })
+            .collect())
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Vec<(Vec<u8>, Vec<u8>)>, IndexedDbContextError> {
+        let mut key_values = vec![];
+        let range = prefix_to_range(key_prefix)?;
+        let transaction = self.database.transaction_on_one(&self.object_store_name)?;
+        let object_store = transaction.object_store(&self.object_store_name)?;
+        let Some(cursor) = object_store.open_cursor_with_range_owned(range)?.await? else {
+            return Ok(key_values);
+        };
+
+        loop {
+            let Some(key) = cursor.primary_key() else {
+                break;
+            };
+            let key = js_sys::Uint8Array::new(&key);
+            key_values.push((
+                key.subarray(key_prefix.len() as u32, key.length()).to_vec(),
+                js_sys::Uint8Array::new(&cursor.value()).to_vec(),
+            ));
+            if !cursor.continue_cursor()?.await? {
+                break;
+            }
+        }
```

Comment:
It is a little bit puzzling that the `find_key_values_by_prefix` implementation is significantly different from the `find_keys_by_prefix`. Is it questions of ownership?

Reply:
It's because the IndexedDB API has a convenient one-shot function for getting a range of keys, or a range of values, but not for getting both keys and values.  Rather than perform one query to get the list of keys and then _n_ queries to get a value for each key, I fell back to the slightly lower-level cursor API, which does allow getting both keys and values.